snuba/datasets/discover.py

```diff
@@ -220,6 +221,19 @@ def __init__(self) -> None:
                 ("version", Nullable(String())),
                 ("http_method", Nullable(String())),
                 ("http_referer", Nullable(String())),
+                ("os_build", Nullable(String())),
+                ("os_kernel_version", Nullable(String())),
+                ("device_name", Nullable(String())),
+                ("device_brand", Nullable(String())),
+                ("device_locale", Nullable(String())),
+                ("device_uuid", Nullable(String())),
+                ("device_model_id", Nullable(String())),
+                ("device_arch", Nullable(String())),
+                ("device_battery_level", Nullable(Float(32))),
+                ("device_orientation", Nullable(String())),
+                ("device_simulator", Nullable(UInt(8))),
+                ("device_online", Nullable(UInt(8))),
+                ("device_charging", Nullable(UInt(8))),
```

Comment:
All these fields are promoted contexts, they should never be queries directly (https://github.com/getsentry/snuba/blob/master/snuba/datasets/events.py#L142-L156) they should always be queried as `contexts['device.name']`. 

Do you have an example of query where these columns are queried directly ?

Reply:
As far as I know nothing in sentry ever referenced a column in the form "contexts[..]"

Reply:
`If all these values are available through contexts in the errors, events and transactions datasets, we could update sentry to use contexts[] instead of the concrete fields.` 

Can I consider this a plan for "we are going to stop referencing directly the promoted columns" :) ?

`We have been using the concrete columns where possible as they generally perform better.`
This is not 100% accurate:
1) if you query a tag/context and that one is promoted (and it is not one of the exceptions in the next point) then there is no performance gain whatsoever since Snuba translates the query into the concrete column. Though using the tags[]/contests[] column make things easier to understand because you do not have to dela with different data types and you do not have to deal with two ways to do the same thing.
2) there are a few columns in transactions/discover where this is not true, like the trace_id because the context is formatted as string and the column is formatted as UUID. We need some additional query processing capability before we are able to avoid this. But that applies only to a handful of columns in transactions.
3) if you have a tags[]/contexts[] in the query and that is not promoted, then you do not even have a concrete column to query.

So, unless you are looking for a transactions specific columns, go for tags/contexts. And even in the case of transactions, we are going to fix it.

Reply:
@lynnagara We are using contexts for a couple transaction fields in sentry. See https://github.com/getsentry/sentry/blob/master/src/sentry/snuba/events.py#L169

@fpacifici So is the ideal case here to switch to using context[...] syntax for all queries against promoted columns? That's easily done I think. How do I tell which columns should be queried using the context/tag?

Reply:
@evanh All of the `device.*`, `sdk.*`, `http.*`, `geo.*`, `os.*` fields come from contexts.

Reply:
My thinking of 'performs better' was that clickhouse performs better using the concrete columns over unpacking the contexts arrays. Given that snuba can upgrade a `contexts[]` field to the concrete column transparently using contexts would be simpler overall.

Reply:
@evanh `context[...] syntax for all queries against promoted columns?` . contexts[...] for promoted contexts and tags[...] for promoted tags. They both exists.

both, your call, let me know if you want to migrate to contexts instead of adding these columns or if you need this as a stop gap.

Reply:
@fpacifici I am almost done with a PR shifting all the fields Mark mentioned to use contexts[...], but which fields are promoted tags and should use tags[...]?